### ISSUE TITLE
Using an HTTPS proxy hangs curl.

### ISSUE THREAD

**Issue by agl** [curl/curl#1156]
HTTPS proxy support was added in cb4e2be7c6d42ca0780f8e0a747cecf9ba45f151.

However, using it appears to cause curl to hang. [This loop](https://github.com/curl/curl/blob/dacfecb7b676b6ace6d707d65fefdadd28301bd4/lib/http_proxy.c#L323) reads the proxy's response byte-by-byte. It loops based on whether the socket to the proxy is readable. However, when the connection to the proxy is using TLS, the TLS library will read a whole record (even when asked for a single byte) and drain the socket. Thus only the first byte of the response is read. After that, the socket is never readable because the rest of the response is waiting in the TLS library's buffer.

[This gist](https://gist.github.com/agl/2da9e61528f0d373894d8c919abfb518) contains a dummy HTTPS proxy (that always connects to Google) that can be used to test this. Once running, try:

`curl --proxy-insecure  --proxy https://127.0.0.1:8080 https://www.google.com -v -v`

In the current git head, that command will hang and `strace` will show it polling the socket to the proxy once a second.

**Comment by bagder** [comment]
Thanks, I probably made things worse for this with the 3ea3518429c00c42 commit...

**Comment by jay** [comment]
can we solve this with ssl pending, it works here. 
~~~diff
diff --git a/lib/http_proxy.c b/lib/http_proxy.c
index fc54741..2e48006 100644
--- a/lib/http_proxy.c
+++ b/lib/http_proxy.c
@@ -311,6 +311,7 @@ CURLcode Curl_proxyCONNECT(struct connectdata *conn,
       perline=0;
 
       while((nread<BUFSIZE) && (keepon && !error)) {
+        int pending;
 
         check = Curl_timeleft(data, NULL, TRUE);
         if(check <= 0) {
@@ -320,7 +321,13 @@ CURLcode Curl_proxyCONNECT(struct connectdata *conn,
         }
 
         /* loop every second at least, less if the timeout is near */
-        switch (SOCKET_READABLE(tunnelsocket, check<1000L?check:1000)) {
+        pending = SOCKET_READABLE(tunnelsocket, 0);
+        if(pending == 0)
+          pending = Curl_ssl_data_pending(conn, FIRSTSOCKET);
+        if(pending == 0)
+          pending = SOCKET_READABLE(tunnelsocket, check<1000L?check:1000);
+
+        switch (pending) {
         case -1: /* select() error, stop reading */
           error = SELECT_ERROR;
           failf(data, "Proxy CONNECT aborted due to select/poll error");
~~~

a few things I'm unclear on, 
- do we have similar exposure where we aggressively call recv on windows to slurp data and we have a cache
- is this check always FIRSTSOCKET
- is socket readable 0 valid, my intention there is basically avoid a situation where ssl_pending always returns true because it has some cached encrypted data but decrypted data may not be available (this is possible in schannel at least) and if that occurs when the socket goes bad we don't want to hang checking for pending data with a bad socket

**Comment by bagder** [comment]
> do we have similar exposure where we aggressively call recv on windows

I don't know. Can you point to specific code where you suspect it can happen?

> is this check always FIRSTSOCKET

No, I believe it should use the `sockindex` argument there as it can be another socket for FTPS.

> is socket readable 0 valid

I think it is, but this is not exactly my backyard territory so I'm not totally sure.

Two additional things:

1. We can't unconditionally ask for pending SSL data since it doesn't have to be done over SSL!
2. There's also a second SOCKET_READABLE() call 20 or so lines above which probably also needs adjusting in a similar way

**Comment by jay** [comment]
>> do we have similar exposure where we aggressively call recv on windows

> I don't know. Can you point to specific code where you suspect it can happen?

no i was musing, i seem to remember we're doing in windows a recv for every send or something so there's a recv buffer

> We can't unconditionally ask for pending SSL data since it doesn't have to be done over SSL!

we do this elsewhere i believe because the ssl_pending macro evaluates to 0 if not built with ssl or the handle is not ssl... unless I misunderstand.

to everything else you wrote ok it's in the second draft, and i adjusted for the progress function so it's called at least once a second
~~~diff
diff --git a/lib/http_proxy.c b/lib/http_proxy.c
index fc54741..9634454 100644
--- a/lib/http_proxy.c
+++ b/lib/http_proxy.c
@@ -285,7 +285,8 @@ CURLcode Curl_proxyCONNECT(struct connectdata *conn,
     }
 
     if(!blocking) {
-      if(0 == SOCKET_READABLE(tunnelsocket, 0))
+      if(0 == SOCKET_READABLE(tunnelsocket, 0) &&
+         !Curl_ssl_data_pending(conn, sockindex))
         /* return so we'll be called again polling-style */
         return CURLE_OK;
       else {
@@ -311,6 +312,10 @@ CURLcode Curl_proxyCONNECT(struct connectdata *conn,
       perline=0;
 
       while((nread<BUFSIZE) && (keepon && !error)) {
+        int pending;
+
+        if(Curl_pgrsUpdate(conn))
+          return CURLE_ABORTED_BY_CALLBACK;
 
         check = Curl_timeleft(data, NULL, TRUE);
         if(check <= 0) {
@@ -319,8 +324,18 @@ CURLcode Curl_proxyCONNECT(struct connectdata *conn,
           break;
         }
 
-        /* loop every second at least, less if the timeout is near */
-        switch (SOCKET_READABLE(tunnelsocket, check<1000L?check:1000)) {
+        /* we wait a total of a second at most for each loop iteration if no
+           data avail. this is because the progress function should be called
+           at the documented interval of at least once a second.
+           the time is split between SOCKET_READABLE (max 900ms) and Curl_read
+           returning CURLE_AGAIN (max 100ms). */
+        pending = SOCKET_READABLE(tunnelsocket, 0);
+        if(pending == 0)
+          pending = Curl_ssl_data_pending(conn, sockindex);
+        if(pending == 0)
+          pending = SOCKET_READABLE(tunnelsocket, check<900L?check:900);
+
+        switch(pending) {
         case -1: /* select() error, stop reading */
           error = SELECT_ERROR;
           failf(data, "Proxy CONNECT aborted due to select/poll error");
@@ -333,8 +348,10 @@ CURLcode Curl_proxyCONNECT(struct connectdata *conn,
             return CURLE_RECV_ERROR;
           }
           result = Curl_read(conn, tunnelsocket, ptr, 1, &gotbytes);
-          if(result==CURLE_AGAIN)
+          if(result == CURLE_AGAIN) {
+            Curl_wait_ms(check < 100 ? (int)check : 100);
             continue; /* go loop yourself */
+          }
           else if(result)
             keepon = FALSE;
           else if(gotbytes <= 0) {
@@ -553,10 +570,11 @@ CURLcode Curl_proxyCONNECT(struct connectdata *conn,
           }
           break;
         } /* switch */
-        if(Curl_pgrsUpdate(conn))
-          return CURLE_ABORTED_BY_CALLBACK;
       } /* while there's buffer left and loop is requested */
 
+      if(Curl_pgrsUpdate(conn))
+        return CURLE_ABORTED_BY_CALLBACK;
+
       if(error)
         return CURLE_RECV_ERROR;
 
~~~

**Comment by jay** [comment]
draft3 at https://github.com/curl/curl/compare/master...jay:fix_proxy_CONNECT_hang?expand=1
I added a function in https_proxy to check for pending data, it also checks for the cached recv data before waiting on the socket
~~~c
static bool conn_has_data_pending(struct connectdata *conn, int sockindex)
{
  return SOCKET_READABLE(conn->sock[sockindex], 0) > 0 ||
         Curl_ssl_data_pending(conn, sockindex) ||
         Curl_recv_has_postponed_data(conn, sockindex);
}
~~~
It may be necessary to make this a Curl_ function if there's other code that does something similar,

**Comment by bagder** [comment]
I think it looks fine (as I assume it also passes the test suite and @agl 's test case)! We really need to get a HTTPS proxy test setup to make sure we don't get more regressions like this going forward...

**Comment by jay** [comment]
on draft4 now, moved conn data pending func into connect as a Curl_conn_data_pending, and I redid the proxy connect so that the socket wait only happens after the read returns eagain. in the traditional compare a lot of lines appear modified but it's mostly only whitespace, the link below suppresses the whitespace changes (w=1)
https://github.com/curl/curl/compare/master...jay:fix_proxy_CONNECT_hang?expand=1&w=1

**Comment by jay** [comment]
I'm pretty much done with this but I have one more question @bagder. When a 407 response body must be ignored the [end-of-headers code](https://github.com/jay/curl/blob/469561b233608527c20689e03792ca0f11eb037f/lib/http_proxy.c#L437-L493) that initially processes the first piece of body sets TUNNEL_COMPLETE regardless of whether all the content is read (in 1-byte-per cases this is unlikely now) but yet in the 'chunk reading DONE' (same highlighted area) it also sets that state. I'm not sure why is it setting it in chunk completed there if it should always be set. And the [block in the loop](https://github.com/jay/curl/blob/469561b233608527c20689e03792ca0f11eb037f/lib/http_proxy.c#L364-L399) that does the rest of the 407 body read sets TUNNEL_COMPLETE in 'chunk reading DONE' but not for example when the content length is fully read. This is also confusing because the initial end-of-header code has set it already in all cases of keepon == 2. This doesn't have to do with my changes and can be seen in [7.51 code](https://github.com/curl/curl/blob/curl-7_51_0/lib/http_proxy.c#L466) as well.

Also: now that it's reading 1-byte-per for the connect there conceivably could be a lot of 
Read 1 bytes of chunk, continue
Read 1 bytes of chunk, continue
Read 1 bytes of chunk, continue
...

**Comment by bagder** [comment]
I think you should consider landing this asap so that we can have this functionality working in the release that I am to build in two days. The `TUNNEL_COMPLETE` thing probably works because we read the response there in a blocking manner anyway so it should be fine to set it early or even many times, I think.

The "Read 1 bytes of chunk, continue" output should just be removed as well as you inferred, that won't make anyone happy.

And then we can of course keep on polishing whatever needs to get improved there.

**Comment by jay** [comment]
Landed in [d00f2a8](https://github.com/curl/curl/commit/d00f2a8f2e8529024709f70c8ab4c4138dc651a0?w=1)

**Comment by bagder** [comment]
As this particular flaw is now believed to be fixed, I'm closing this!
